File: methods/survival/templates/R/competing_risks.R
#!/usr/bin/env Rscript
# Template: Aalen–Johansen / Fine–Gray vs 1−KM that censors the competing event.
# Requires cause: 0 = censored, 1 = event of interest, 2+ = competing.
# Most GEO ICI series do not deposit this field.

suppressPackageStartupMessages({
  library(survival)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("usage: competing_risks.R input.csv horizon [time_col cause_col]")
}
infile <- args[[1]]
horizon <- as.numeric(args[[2]])
time_col <- if (length(args) >= 3) args[[3]] else "time"
cause_col <- if (length(args) >= 4) args[[4]] else "cause"

df <- read.csv(infile, stringsAsFactors = FALSE)
df <- df[complete.cases(df[, c(time_col, cause_col)]), ]
causes <- sort(unique(df[[cause_col]]))
if (all(causes %in% c(0, 1))) {
  stop("no competing event is coded. Do not run a competing-risks model just because the playbook mentions one.")
}

# 1-KM that incorrectly censors competing events
km <- survfit(Surv(df[[time_col]], as.integer(df[[cause_col]] == 1)) ~ 1)
km_at <- summary(km, times = horizon, extend = TRUE)
naive <- 1 - km_at$surv

# Aalen-Johansen via survfit(Surv(..., type = "mstate"))
# status must be a factor whose first level is censoring
st <- factor(df[[cause_col]], levels = sort(unique(c(0, df[[cause_col]]))))
aj <- survfit(Surv(df[[time_col]], st, type = "mstate") ~ 1)
aj_sum <- summary(aj, times = horizon, extend = TRUE)
# pstate columns follow the factor levels excluding the censoring level
cif <- NA_real_
if (!is.null(aj_sum$pstate)) {
  # column for cause 1
  cause_levels <- levels(st)[-1]
  idx <- match("1", cause_levels)
  if (!is.na(idx)) cif <- aj_sum$pstate[1, idx]
}

cat(sprintf("Aalen-Johansen CIF(cause=1, t=%g) = %.3f\n", horizon, cif))
cat(sprintf("1-KM treating competing events as censoring     = %.3f\n", naive))
cat(sprintf("absolute overestimate                           = %.3f\n", naive - cif))

if (requireNamespace("cmprsk", quietly = TRUE) && "marker" %in% names(df)) {
  cat("\nFine-Gray on a binary high/low marker (illustrative):\n")
  fg <- cmprsk::crr(
    ftime = df[[time_col]],
    fstatus = df[[cause_col]],
    cov1 = data.frame(high = as.integer(df$marker > median(df$marker, na.rm = TRUE))),
    failcode = 1,
    cencode = 0
  )
  print(summary(fg))
}

File: methods/survival/templates/python/competing_risks.py
#!/usr/bin/env python3
"""Template: Aalen–Johansen CIF vs 1−KM that censors the competing event (lifelines).

Use when the event of interest (progression, irAE, TTR) can be precluded by
death or treatment switch. Requires a cause code: 0 = censored, 1 = event of
interest, 2+ = competing events. Most GEO ICI series do not deposit this field.
"""

from __future__ import annotations

import argparse
import sys

import pandas as pd
from lifelines import AalenJohansenFitter, KaplanMeierFitter


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--input", required=True)
    ap.add_argument("--time-col", default="time")
    ap.add_argument("--cause-col", default="cause", help="0=censored, 1=interest, 2+=competing")
    ap.add_argument("--horizon", type=float, required=True)
    args = ap.parse_args()

    df = pd.read_csv(args.input)[[args.time_col, args.cause_col]].dropna()
    causes = set(df[args.cause_col].astype(int))
    if causes <= {0, 1}:
        sys.exit(
            "no competing event is coded. Do not run a competing-risks model "
            "just because the playbook mentions one."
        )
    time = df[args.time_col]
    cause = df[args.cause_col].astype(int)
    aj = AalenJohansenFitter(calculate_variance=True)
    aj.fit(time, cause, event_of_interest=1)
    cif = float(aj.predict(args.horizon))
    km = KaplanMeierFitter().fit(time, (cause == 1).astype(int))
    naive = float(1 - km.predict(args.horizon))
    print(f"Aalen-Johansen CIF(cause=1, t={args.horizon:g}) = {cif:.3f}")
    print(f"1-KM treating competing events as censoring     = {naive:.3f}")
    print(f"absolute overestimate                           = {naive - cif:.3f}")


if __name__ == "__main__":
    main()
